index.js

```diff
@@ -99,7 +94,60 @@ async function pay (plugin, {
   }
 }
 
+async function pull (plugin, {
+  pointer,
+  amount,
+  streamOpts = {},
+  callback,
+  callbackOpts = {}
+}) {
+  await plugin.connect()
+  const receiveMax = amount || Infinity
+
+  let response
+  try {
+    response = await query(pointer)
+  } catch (err) {
+    const endpoint = new URL(pointer.startsWith('$')
+      ? 'https://' + pointer.substring(1)
+      : pointer)
+    const postResponse = await fetch(endpoint.href, {
+      method: 'post'
+    })
+    if (postResponse.status !== 200) {
+      throw new Error('got error response from spsp server.' +
+        ' endpoint="' + endpoint.href + '"' +
+        ' status=' + postResponse.status +
+        ' message="' + (await postResponse.text()) + '"')
+    }
+  }
+  response = await query(pointer)
+  if (response.contentType.indexOf('application/spsp4+json') !== -1) {
+    const ilpConn = await createConnection({
+      plugin,
+      destinationAccount: response.destinationAccount,
+      sharedSecret: response.sharedSecret,
+      ...streamOpts
+    })
+
+    const stream = await ilpConn.createStream()
+
+    try {
+      await stream.receiveTotal(receiveMax)
+      callback(receiveMax, callbackOpts)
+    } catch (err) {
+      const receivedAmount = Number(err.message.match(/totalReceived:\s([0-9]*)/)[1])
+      callback(receivedAmount, callbackOpts)
+    }
+    return new Promise(resolve => ilpConn.on('end', resolve))
```

Comment:
Why does this wait for the connection to end, whereas `pay` waits for the stream to end?

Reply:
Ah, I overlooked the `Promise.race` in `pay`. Either way, why do we need to wait for the connection (or even the stream) to `end` here?
I'm trying to think of how this could return the received amount (and `pay` could return the total sent amount) without using a callback.

Reply:
We do not have to wait until the connection ends, the stream would be sufficient. However, if we do not wait for at least the stream to end, then you cannot be certain about the actual amount sent (in push) or received (in pull).

Reply:
It sounds like the best thing to do here is 
```suggestion
    return await ilpConn.end()
```
https://github.com/interledgerjs/ilp-protocol-stream/issues/108#issuecomment-466561170

Reply:
I think this change to the server now makes it ok to do `await ilpConn.end()` here
https://github.com/sabinebertram/ilp-spsp-pull-payment-server/commit/bf4afd2a1a060fae71cfa61768d97a54d66fcfb7

Reply:
The connection is ended [here](https://github.com/sabinebertram/ilp-protocol-spsp/blob/7ed33bd79bdf5585975cc00207633c35ec02367f/index.js#L141). Are you saying that we should wait for it to end or that you don't want to return the Promise?

Reply:
Since it actually works now, I'm saying
```
    ilpConn.end()
    return new Promise(resolve => ilpConn.on('end', resolve))
```
can just be combined to be
```
    await ilpConn.end()```
I think either way, we're waiting for the connection to end.